controller/src/main/java/io/pravega/controller/server/SegmentStoreConnectionManager.java

```diff
@@ -0,0 +1,653 @@
+                        waiting.getFuture().complete(connectionObject);
+                    } else {
+                        if (!isRunning) {
+                            // The connection will be closed if returned anytime after the shutdown has been initiated.
+                            log.debug("ConnectionManager is shutdown");
+                            connectionCount--;
+                            toClose = true;
+                        } else {
+                            // as connections are returned to us, we put them in queue to be reused
+                            if (availableConnections.size() < maxAvailableConnections) {
+                                // if returned connection increases our available connection count, do not include it
+                                log.debug("Returned connection object is included in the available list");
+                                availableConnections.offer(connectionObject);
+                            } else {
+                                log.debug("Returned connection object is discarded as available list is full");
+                                connectionCount--;
+                                toClose = true;
+                            }
+                        }
+                    }
+                }
+
+                if (toClose) {
+                    if (connectionListener != null) {
+                        connectionListener.notify(ConnectionListener.ConnectionEvent.ConnectionClosed);
+                    }
+                    connectionObject.connection.close();
+                }
+            }
+        }
+
+        /**
+         * This method will try to create a new connection only if BOTH conditions are met
+         * 1. connectionCount is less than MAX_CONCURRENT_CONNECTIONS.
+         * 2. it has taken a waiting request from the waitQueue.
+         */
+        private void tryCreateNewConnection() {
+            WaitingRequest waiting = null;
+            synchronized (lock) {
+                if (connectionCount < maxConcurrentConnections) {
+                    waiting = waitQueue.poll();
+                    if (waiting != null) {
+                        log.debug("Creating new connection for {}. Total connection count = {}", uri, connectionCount);
+                        connectionCount++;
+                    }
+                }
+            }
+
+            if (waiting != null) {
+                log.debug("Creating new connection for {}", uri);
+                ReusableReplyProcessor rp = new ReusableReplyProcessor();
+                rp.initialize(waiting.getReplyProcessor());
+                Futures.completeAfter(
+                        () -> clientCF.establishConnection(uri, rp)
+                                      .thenApply(connection -> new ConnectionObject(connection, rp))
+                                      .whenComplete((r, e) -> {
+                                          if (connectionListener != null) {
```

Comment:
What if an error occurred? Do you still want to notify of a new connection?

Also, `thenApply` and `whenComplete` can be merged together.

Reply:
since we assigned this new connection creation to a request, that request will get the failure and will retry after returning the failed connection to this class.